### PR TITLE
Add network address dimension, group, and group membership tables

### PR DESCRIPTION
**Pull request by talanknight**
This also creates a function to generate new groups when the set of addresses associated with an existing group is not captured in an existing group and updates the wh_upsert_host function to use that function.

Future PRs will include 
1. migrating existing data in the data warehouse from the host dimension table to these newly created tables and 
2. allowing static hosts to have multiple ip addresses and/or dns names.

### CODE REVIEW

**Review by tmessi — CHANGES_REQUESTED** [review]
Can you add some additional tests that are similar to the ones for credential dimension:

- [three_credentials_one_change.sql](https://github.com/hashicorp/boundary/blob/alanknight_dw_plugin_hosts/internal/db/sqltest/tests/wh/credential_dimension/three_credentials_one_change.sql)
- [session_multiple_sessions.sql](https://github.com/hashicorp/boundary/blob/alanknight_dw_plugin_hosts/internal/db/sqltest/tests/wh/credential_dimension/session_multiple_sessions.sql)
- [session_update.sql](https://github.com/hashicorp/boundary/blob/alanknight_dw_plugin_hosts/internal/db/sqltest/tests/wh/credential_dimension/session_update.sql)

The tests should cover scenarios where one ore more session is created for a host, with the host's addresses being updated in different ways between each session.

**Comment by talanknight** [comment]
I've added to our existing tests/wh/host_dimesnsion/session_multiple_sessions.sql test for the case of modifying the set of addresses associated with a host and creating another session or modifying an existing session.  I don't think tthree_credentials_one_change.sql has a related applicable test for this change though.

**Review by tmessi — APPROVED** [review]

**Review by mgaffney — CHANGES_REQUESTED** [review]
Initial comments. I still have more to review.

File: internal/db/schema/migrations/oss/postgres/24/02_wh_network_address_dimensions.up.sql [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  -- wh_private_address_status returns a warehouse appropriate string
+  -- representing if the address is private or public.
+  create function wh_private_address_status(inet) returns text
+  as $$
+  begin
+    case
+      when $1 << any ('{10.0.0.0/8, 192.168.0.0/16, 172.16.0.0/12, fc00::/7, fe80::/10}'::cidr[]) then
+        return 'Private';
+      else
+        return 'Public';
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/db/schema/migrations/oss/postgres/24/02_wh_network_address_dimensions.up.sql]: I'll change it to 'Not Applicable'.  There is a test case that covers this in `tests/wh/network_address_dimension/helper_functions.sql` which showed it as 'Public' which clearly isn't right.

File: internal/db/schema/migrations/oss/postgres/24/02_wh_network_address_dimensions.up.sql [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+      end case;
+  end;
+  $$ language plpgsql;
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/db/schema/migrations/oss/postgres/24/02_wh_network_address_dimensions.up.sql]: It doesn't return null on null.  Updated to add 'immutable'.

File: internal/db/schema/migrations/oss/postgres/24/03_wh_network_address_dimensions.up.sql [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+      for a_key in
+        select address
+        from whx_network_address_dimension_source
+        where host_id = p_host_id
+        loop
+          insert into wh_network_address_group_membership
+          (network_address_group_key, network_address)
+          values
+            (nag_key,               a_key);
+        end loop;
```
Comment (mgaffney) [inline internal/db/schema/migrations/oss/postgres/24/03_wh_network_address_dimensions.up.sql:70]: Can this loop be replaced with a single insert statement? I tried this and the tests passed:

```sql
      insert into wh_network_address_group_membership
        (network_address_group_key, network_address)
      select nag_key, address
        from whx_network_address_dimension_source
        where host_id = p_host_id;
```

**Review by mgaffney — CHANGES_REQUESTED** [review]

File: internal/db/schema/migrations/oss/postgres/24/02_wh_network_address_dimensions.up.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/db/schema/migrations/oss/postgres/24/02_wh_network_address_dimensions.up.sql]: Good point.  Made it return null on null input.

File: internal/db/sqltest/initdb.d/01_colors_persona.sql [hunk 4]
```diff
@@ -145,6 +145,52 @@ begin;
...
+    ('h_____cr__06', '6.red.color'),
+    ('h_____cr__07', '7.red.color'),
+    ('h_____cr__08', '8.red.color'),
+    ('h_____cr__09', '9.red.color');
+
+  insert into host_ip_address
+    (host_id, address)
+  values
+    -- note there are no address for blue 1
+    ('h_____cb__02', '10.0.0.2'),
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/db/sqltest/initdb.d/01_colors_persona.sql:175]: Done.

File: internal/db/schema/migrations/oss/postgres/24/02_wh_network_address_dimensions.up.sql [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+      when $1 is null then
+        return 'Not Applicable';
+      else
+        return 'Public IP address';
+      end case;
+  end;
+  $$ language plpgsql
+    immutable;
+
+  create view whx_network_address_dimension_source as
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/db/schema/migrations/oss/postgres/24/02_wh_network_address_dimensions.up.sql:55]: Done.

File: internal/db/schema/migrations/oss/postgres/24/03_wh_network_address_dimensions.up.sql (hunk 3 above)
Comment (talanknight) [reply internal/db/schema/migrations/oss/postgres/24/03_wh_network_address_dimensions.up.sql]: That works great! Thanks!

**Review by mgaffney — APPROVED** [review]
Nice work!

**Pull request merged**